lib/boombox.ex

```diff
@@ -125,8 +125,66 @@ defmodule Boombox do
           input: input(),
           output: output()
         ) :: :ok | Enumerable.t()
-  @endpoint_opts [:input, :output]
   def run(stream \\ nil, opts) do
+    opts = validate_opts!(stream, opts)
+
+    case opts do
+      %{input: {:stream, _stream_opts}} ->
+        procs = start_pipeline(opts)
+        consume_stream(stream, procs)
+
+      %{output: {:stream, _stream_opts}} ->
+        produce_stream(opts)
+
+      opts ->
+        opts
+        |> start_pipeline()
+        |> await_pipeline()
+    end
+  end
+
+  @doc """
+  Asynchronous version of run/2
+  Doesn't block the calling process until the termination of the pipeline.
+  It returns a `Task` that can be awaited later.
+  If the output is a stream the behaviour is idential to run/2
+  """
+  @spec async(Enumerable.t() | nil,
+          input: input(),
+          output: output()
+        ) :: Task.t() | Enumerable.t()
+  def async(stream \\ nil, opts) do
+    opts = validate_opts!(stream, opts)
+
+    case opts do
+      %{input: {:stream, _stream_opts}} ->
+        procs = start_pipeline(opts)
+
+        Task.async(fn ->
+          Process.monitor(procs.supervisor)
+          consume_stream(stream, procs)
+        end)
+
+      %{output: {:stream, _stream_opts}} ->
+        produce_stream(opts)
```

Comment:
[NIT] Since `start_pipeline` is removed from `consume_stream`, I believe it might be worth to remove it from `produce_stream` as well, for the sake of consistency